File: january/19/merge sort.py
def merge(left,right):
    result=[]
    i,j=0,0
    while i<len(left) and j<len(right):
        if left[i]<right[j]:
            result.append(left[i])
            i+=1
        else:
            result.append(right[j])
            j+=1
    while i<len(left):
        result.append(left[i])
        i+=1
    while j<len(right):
        result.append(right[j])
        j+=1
    return result

def merge_sort(list):
    if len(list)<2:
        return list[:]
    else:
        middle=len(list)//2
        left=merge_sort(list[:middle])
        right=merge_sort(list[middle:])
        return merge(left,right)
    
print(merge_sort([9,5,1,7,6,8,4]))

File: february/12/merge_sort.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<int> merge(vector<int> left,vector<int> right){
    int i{};
    int j{};
    vector<int> result;
    while (i<left.size() && j<right.size()){
        if (left[i]<right[j]){
            result.push_back(left[i]);
            i++;
        }
        else{
            result.push_back(right[j]);
            j++;
        }
    }
    while (i<left.size()){
        result.push_back(left[i]);
        i++;
    }
    while(j<right.size()){
        result.push_back(right[j]);
        j++;
    }
    return result;
}

vector<int> merge_sort(vector<int> arr){
    if (arr.size()<2){
        return arr;
    }

    else{
        vector<int> arr_left;
        vector<int> arr_right;
        int mid = arr.size()/2;

        for(int i=0;i<mid;i++){
            arr_left.push_back(arr[i]);
        }

        for(int i=mid;i<arr.size();i++){
            arr_right.push_back(arr[i]);
        }

        vector<int> left=merge_sort(arr_left);
        vector <int> right=merge_sort(arr_right);
        
        return merge(left,right);
    }
}

int main(){
    vector<int> arr{38,27,43,3,9,82,10};
    vector<int> sorted_arr=merge_sort(arr);
    for (int i=0;i<sorted_arr.size();i++){
        cout<<sorted_arr[i]<<" ";
    }
    cout<<endl;

    return 0;
}